Google Ads: Navigation from Hub menu entry point

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13218 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR handles the navigation for GLA from the Hub menu entry point. Changes include:
- Updated `DefaultGoogleAdsEligibilityChecker` to optimize the check following the technical post pe5sF9-2O6-p2.
- Updated `HubMenuViewModel` to check if the site has campaigns to determine the web page to display:
  - If the site has existing campaigns, display the dashboard page.
  - Otherwise, display the campaign creation page.
- Updated `WebView` and `AuthenticatedWebView` to support `exitTrigger` without `urlToTriggerExit`.
- Added `GoogleAdsCampaignCoordinator` to handle navigation to Google ads campaigns.
- Updated `HubMenu` and `HubMenuHostingController` to start the coordinator when tapping the Google ads entry point.

## Testing steps
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Pre-requisite: Install, activate, and set up the Google Ads plugin to a test site following pe5sF9-2Qo-p2. Ensure that you're using the plugin version 2.7.5 or above.
- Log in to your test store and navigate to the Menu tab.
- Select Google for WooCommerce. Confirm that a web view is presented for campaign management.
  - If the store is a self-hosted store authenticated with WPCom credentials, you'd be asked to log in with WPOrg credentials. 
  - If the store is a WPCom store or self-hosted store authenticated with WPOrg credentials, you'd be authenticated automatically.
- Assuming your site has no paid campaign created yet, tapping the entry point should display the creation flow. Proceed to create a test campaign.
- When the campaign is created successfully, the web view should be dismissed.
- Tap on the entry point again, this time you should see the campaign dashboard page.


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/5533851/209da385-6a52-4c53-a56d-1dbb4bba8927



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Optimize eligibility check for google ads

itsmeichigo committed:
Check if site has campaign on hub menu

itsmeichigo committed:
Support exit trigger handling for WebView

itsmeichigo committed:
Support redirects for GLA campaign management

itsmeichigo committed:
Handle tap on GLA item on hub menu

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Add localized string for web view title

itsmeichigo committed:
Fix test failures

itsmeichigo committed:
Display google ads in a sheet instead of detail view

itsmeichigo committed:
Update campaign check after creation success

itsmeichigo committed:
Update tests for HubMenuViewModel

itsmeichigo committed:
Revert changes to HubMenu and HubMenuViewModel

itsmeichigo committed:
Add hosting controller for WebView

itsmeichigo committed:
Add GoogleAdsCampaignCoordinator

itsmeichigo committed:
Start google ads campaign coordinator from HubMenuViewController

itsmeichigo committed:
Remove irrelevant unit tests for HubMenuViewModel

itsmeichigo committed:
Fix swiftlint violation

itsmeichigo committed:
Revert change to HubMenuViewModel

itsmeichigo committed:
Add tests for GoogleAdsCampaignCoordinator

itsmeichigo commented:
@selanthiraiyan To make the logic reusable for the new dashboard card later, I reverted most changes to `HubMenuViewModel` and `HubMenu` and added `GoogleAdsCampaignCoordinator`. The Google ads campaign view is now presented using UIKit instead. This PR is ready for another look, thanks!

itsmeichigo committed:
Revert unnecessary changes

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/WebView.swift:
```diff
@@ -71,9 +100,27 @@ struct WebView: UIViewRepresentable {
+                    if let urlToTriggerExit = parent.urlToTriggerExit,
+                        url.absoluteString.contains(urlToTriggerExit) {
+                        parent.exitTrigger?(url)
+                    } else {
+                        parent.exitTrigger?(url)
+                    }
```
Could you please explain why we are calling `exitTrigger?(url)` in both if and else conditions? 

Is the idea to fire the `exitTrigger` only if the `url` matches when `urlToTriggerExit` is specified? I think the `else` flow will be executed if there is no match found. This way, we will end up calling `exitTrigger` for all published `url` change events. 🤔

itsmeichigo replied:
This change lets the caller decide if the web view should "exit" when `urlToTriggerExit` is not specified. I'm good with changing the name of this closure to something more generic like `exitHandler`.

itsmeichigo replied:
If you take a closer look at the coordinator, it doesn't specify `urlToTriggerExit`. I figured the query param may not be fixed for the redirect URL, so letting the coordinator check for the query param itself would be safer.

selanthiraiyan replied:
> If you take a closer look at the coordinator, it doesn't specify urlToTriggerExit. I figured the query param may not be fixed for the redirect URL, so letting the coordinator check for the query param itself would be safer.

Thanks for explaining, @itsmeichigo! I think this change will break other places that use both `urlToTriggerExit` and `exitTrigger`. Attaching a recording of Blaze add payment flow, for example. Please note that the web view is dismissed as soon as it is presented. Kindly let me know your thoughts. 


https://github.com/woocommerce/woocommerce-ios/assets/524475/334be0e6-7617-4073-8bcf-6c1143994383

itsmeichigo replied:
@selanthiraiyan Nice catch! I missed this case.. I added an update in b8e605afd33e855f812008657556a72dc33ca7ec to fix the issue above. I also renamed `exitTrigger` to `redirectHandler` to be more generic. Let me know what you think!

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/WebView.swift:
```diff
@@ -14,33 +36,40 @@ struct WebView: UIViewRepresentable {
+    private let disableLinkClicking: Bool
+
+    let urlToTriggerExit: String?
+    let exitTrigger: ((URL?) -> Void)?
```
Kindly consider adding comments to explain these values.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/WebView.swift:
```diff
@@ -14,33 +36,40 @@ struct WebView: UIViewRepresentable {
     }
 
-    private let webView = WKWebView()
+    let webView = WKWebView()
```
❓ Can this be kept as `private`?

itsmeichigo replied:
I intentionally made this internal so that the coordinator can access the property.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/AuthenticatedWebView.swift:
```diff
@@ -24,10 +24,14 @@ final class DefaultAuthenticatedWebViewModel: AuthenticatedWebViewModel {
-            let url,
             url.absoluteString.contains(urlToTriggerExit) {
             exitTrigger?(url)
+        } else {
+            exitTrigger?(url)
```
Same question as in `WebView.swift` about calling `exitTrigger` from both if and else conditions.

itsmeichigo replied:
Same answer as above.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenu.swift:
```diff
@@ -53,7 +56,11 @@ struct HubMenu: View {
         }
 
-        viewModel.selectedMenuID = menu.id
+        if menu.id == HubMenuViewModel.GoogleAds.id {
+            googleAdsCampaignHandler()
+        } else {
```
Suggestion: Instead of adding an if else condition, would using a guard to exit early after tracking `hubMenuOptionTapped` make it more readable?

itsmeichigo replied:
I think both looks the same 🤔 There's not much logic after the `guard` check so I don't see much improvement when using it.

selanthiraiyan commented on WooCommerce/Classes/GoogleAds/GoogleAdsCampaignCoordinator.swift:
```diff
@@ -0,0 +1,108 @@
+            navigationController.dismiss(animated: true)
+
+            // TODO: show success bottom sheet
+            DDLogDebug("🎉 Campaign creation success")
```
Super nit: We could add "Google" to differentiate it from a Blaze campaign.

selanthiraiyan reviewed (commented):
Thanks for refactoring, Huong! Nice work adding a reusable coordinator. 👏 

The Google campaign creation and auto dismissal work as described. 🎉 

I left a question about triggering exit callback from the web views. Kindly take a look. Thanks!

itsmeichigo commented:
Thank you @selanthiraiyan for the review! I responded to your questions, let me know what you think.

itsmeichigo committed:
Update AuthenticatedWebView and WebView logic for redirectHandler

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/212">19.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan reviewed (approved):
Thanks for the updates, Huong! This LGTM now. 🚢
